odoo/orm/domains.py

```diff
@@ -648,7 +648,7 @@ def _optimize_step(self, model: BaseModel, level: OptimizationLevel) -> Domain:
         size = len(children)
         if size > 1:
             # sort children in order to ease their grouping by field and operator
-            children.sort(key=_optimize_nary_sort_key)
+            children = sorted(children, key=_optimize_nary_sort_key)
```

Comment:
Maybe sorting the list in-place can be faster when almost sorted... and avoids creating a new list.
Simply revert the changes on `_flatten()` to always return a list.